tools/rpctypegen/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+extern crate proc_macro;
+extern crate proc_macro2;
+
+use proc_macro::TokenStream;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use syn::{
+    parse_macro_input, Data, DataEnum, DataStruct, DeriveInput, Fields, FieldsNamed, FieldsUnnamed,
+    Lit, Meta, MetaNameValue,
+};
+
+#[cfg(feature = "dump_errors_schema")]
+fn merge(a: &mut Value, b: &Value) {
+    match (a, b) {
+        (&mut Value::Object(ref mut a), &Value::Object(ref b)) => {
+            for (k, v) in b {
+                merge(a.entry(k.clone()).or_insert(Value::Null), v);
+            }
+        }
+        (a, b) => {
+            *a = b.clone();
+        }
+    }
+}
+
+#[derive(Default, Debug, Deserialize, Serialize)]
+struct Schema {
+    pub schema: BTreeMap<String, ErrorType>,
+}
+
+#[cfg(feature = "dump_errors_schema")]
+impl Drop for Schema {
+    fn drop(&mut self) {
+        // std::env::var("CARGO_TARGET_DIR") doesn't exists
+        let filename = "./target/rpc_errors_schema.json";
+        let schema_json = serde_json::to_value(self).expect("Schema serialize failed");
+        let new_schema_json = if let Ok(data) = std::fs::read(filename) {
+            // merge to the existing file
+            let mut existing_schema = serde_json::from_slice::<Value>(&data)
+                .expect("cannot deserialize target/existing_schema.json");
+            merge(&mut existing_schema, &schema_json);
+            existing_schema
+        } else {
+            schema_json
+        };
+        let new_schema_json_string = serde_json::to_string_pretty(&new_schema_json)
+            .expect("error schema serialization failed");
+        std::fs::write(filename, new_schema_json_string)
+            .expect("Unable to save the errors schema file");
+    }
+}
+
+#[derive(Default, Debug, Deserialize, Serialize)]
+struct ErrorType {
+    /// A type name of the error
+    pub name: String,
+    /// Names of subtypes of the error
+    pub subtypes: Vec<String>,
+    /// An error input name and a type
+    pub props: BTreeMap<String, String>,
+}
+
+thread_local!(static SCHEMA: RefCell<Schema> = RefCell::new(Schema::default()));
```

Comment:
We should make sure procedural macroses are never executed by different threads.

Reply:
`rustc` is a single-threaded AFAIK. I guess we are about to came up with the more robust solution in https://github.com/nearprotocol/nearcore/issues/1972 anyway.

Reply:
It definitely runs parallelizes compilation of the crates across the processes or threads (I don't remember exactly, but you can see it for yourself in `htop` when you compile nearcore) so this schema is not going to be reused across the crates.

Reply:
I meant `rustc` for a single crate is single-threded, but not an entire build process, sure. That's why it merges into the json file. Of course it's far from ideal: there is a small chance that it will rewrite the file. Since it's already a hack (and will be rewritten at some point) and since it's not a runtime code I don't think it worth to make it perfect. If you have a better idea at hand - I'd be glad to apply it!